File: leetcode/C/TwoSum.c
#include <stdio.h>
#include <stdlib.h>

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int *twoSum(int *nums, int numsSize, int target,int *returnSize)
{
    returnSize = (int *)malloc(2 * sizeof(int));
    for (int i = 0; i < numsSize; i++)
    {
        int num=target-nums[i];
        for (int j = i+1; j < numsSize; j++)
        {
            if (num==nums[j])
            {
                returnSize[0] = i;
                returnSize[1] = j;
                break;
            }
        }
    }
    return returnSize;
}

int main(int argc,char **argv,char **envp){
    int numsSize;
    while (scanf("%d", &numsSize) != EOF)
    {
        int *nums=(int*)malloc(2*sizeof(int));
        for (int i = 0; i < numsSize; i++)
        {
            scanf("%d",&nums[i]);
        }
        int target,*res;
        scanf("%d",&target);
        int *result=twoSum(nums,numsSize,target,res);
        printf("%d,%d\n", result[0], result[1]);
        free(res);
    }
    return 0;
}

File: leetcode/Golang/TwoSum.go
package main

func twoSum1(nums []int, target int) []int {
	length := len(nums)
	res := make([]int, 0)
	for i := 0; i < length; i++ {
		num := target - nums[i]
		for j := i + 1; j < length; j++ {
			if num == nums[j] {
				res = append(res, i)
				res = append(res, j)
				break
			}
		}
	}
	return res
}

/*
 * golang中map[int]int默认值为0,value用interface好判断
 */
func twoSum2(nums []int, target int) []int {
	length := len(nums)
	res := make([]int, 0)
	m := make(map[int]interface{})
	for i := 0; i < length; i++ {
		num := target - nums[i]
		if m[num] != nil {
			res = append(res, m[num].(int))
			res = append(res, i)
			break
		}
		m[nums[i]] = i
	}
	return res
}

/*
func main() {
	var numSize, num, target int
	for {
		nums := make([]int, 0)
		n, _ := fmt.Scanf("%d", &numSize)
		if n != 1 {
			break
		}
		for i := 0; i < numSize; i++ {
			_, _ = fmt.Scanf("%d", &num)
			nums = append(nums, num)
		}
		_, _ = fmt.Scanf("%d", &target)
		// res := twoSum1(nums, target)
		res := twoSum2(nums, target)
		fmt.Printf("%d,%d\n", res[0], res[1])
	}
}
*/
